### PR TITLE
Add versions to hot-patched symbols

### LINKED ISSUE #420: Cannot scope PHPUnit 8.5
I wanted to release PHPUnit 8.5.9 from PHPUnit's `8.5` branch just now but ran into the following issue:
```
    ____  __  ______     _____
   / __ \/ / / / __ \   / ___/_________  ____  ___  _____
  / /_/ / /_/ / /_/ /   \__ \/ ___/ __ \/ __ \/ _ \/ ___/
 / ____/ __  / ____/   ___/ / /__/ /_/ / /_/ /  __/ /
/_/   /_/ /_/_/       /____/\___/\____/ .___/\___/_/
                                     /_/

PHP Scoper version 0.13.5 2020-09-16 13:19:21 UTC

   0/683 [░░░░░░░░░░░░░░░░░░░░░░░░░░░░]   0%
  69/683 [▓▓░░░░░░░░░░░░░░░░░░░░░░░░░░]  10%
 137/683 [▓▓▓▓▓░░░░░░░░░░░░░░░░░░░░░░░]  20%
 205/683 [▓▓▓▓▓▓▓▓░░░░░░░░░░░░░░░░░░░░]  30%
 274/683 [▓▓▓▓▓▓▓▓▓▓▓░░░░░░░░░░░░░░░░░]  40%
 342/683 [▓▓▓▓▓▓▓▓▓▓▓▓▓▓░░░░░░░░░░░░░░]  50%


 // Memory usage: 9.42MB (peak: 20.29MB), time: 1.79s                           

 410/683 [▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓░░░░░░░░░░░░]  60%
 683/683 [▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓] 100%
In AddPrefixCommand.php line 103:
                                                                               
  Could not parse the file "/usr/local/src/phpunit/build/tmp/phar/phpunit/Fra  
  mework/MockObject/Builder/ParametersMatch.php".                              
                                                                               

In ParserAbstract.php line 158:
                                               
  Syntax error, unexpected T_MATCH on line 15  
                                               

add-prefix [-d|--working-dir WORKING-DIR] [-p|--prefix PREFIX] [-o|--output-dir OUTPUT-DIR] [-f|--force] [-s|--stop-on-failure] [-c|--config CONFIG] [--no-config] [--] [<paths>...]

Result: 1
```
This is line 15 of `ParametersMatch.php`:
```php
interface ParametersMatch extends Match
```
`match` is a reserved word in PHP 8. PHPUnit 8.5 is not supported on PHP 8. I am not using PHP 8 to run PHP-Scoper. I expect to be able to package code that contains syntax that is valid for PHP 7 but not for PHP 8.

### PR DESCRIPTION
**Pull request by smoench**
fixes #420

### CODE REVIEW

File: src/Reflector.php [hunk 1]
```diff
@@ -14,68 +14,36 @@
...
-use JetBrains\PHPStormStub\PhpStormStubsMap;
 use function strtolower;
 
 /**
  * @private
  */
 final class Reflector
 {
-    private const MISSING_CLASSES = [
-        // https://github.com/JetBrains/phpstorm-stubs/pull/594
```
Comment (smoench) [inline src/Reflector.php:29]: Has been merged and is included now

File: src/Reflector.php [hunk 2]
```diff
@@ -87,18 +87,28 @@ final class Reflector
...
      */
     private static function initSymbolList(?array &$symbols, array $source, array $missingSymbols): void
     {
         if (null !== $symbols) {
             return;
         }
 
+        $excludingSymbols = array_keys(
+            array_filter($missingSymbols, static function ($version) {
+                var_dump(PHP_VERSION_ID <= $version);
```
Comment (villfa) [inline src/Reflector.php:100]: to remove

**Comment by smoench** [comment]
It is related to https://github.com/nikic/PHP-Parser/issues/715. Using a none emulative lexer would fix this bug. Specs needs somehow to be skipped on none supported php versions. Any ideas?

File: .github/workflows/build.yml [hunk 3]
```diff
@@ -29,6 +30,11 @@ jobs:
                         coverage: "none"
                         publish-phar: false
                         git-fetch-depth: 1
+                    -   php: "8.0"
```
Comment (smoench) [inline .github/workflows/build.yml:33]: Running phpstan on php 8 fails as the version is not compatible.

**Comment by theofidry** [comment]
>Specs needs somehow to be skipped on none supported php versions. Any ideas?

I think it would make sense to be able to declare the min PHP version required for a given spec or spec file

File: .github/workflows/build.yml (hunk 3 above)
Comment (smoench) [reply .github/workflows/build.yml]: Removed and waiting for #427

File: src/Reflector.php [hunk 4]
```diff
@@ -14,10 +14,10 @@
 
 namespace Humbug\PhpScoper;
 
+use JetBrains\PHPStormStub\PhpStormStubsMap;
```
Comment (smoench) [inline src/Reflector.php:17]: I think StyleCI is wrong

**Comment by smoench** [comment]
I added a `minPhpVersin`-key to the meta spec

File: src/Reflector.php (hunk 4 above)
Comment (villfa) [reply src/Reflector.php:17]: It applies the `ordered_use` fixer: https://docs.styleci.io/fixers#alpha_ordered_imports

**Comment by smoench** [comment]
Ready for review. In #420 Sebastian Bergmann has proved it is working.

File: src/Reflector.php [hunk 5]
```diff
@@ -87,18 +87,27 @@ final class Reflector
...
+            array_filter($missingSymbols, static function ($version) {
+                return PHP_VERSION_ID < $version;
+            })
+        );
+
         $symbols = array_fill_keys(
-            array_merge(
-                array_keys($source),
-                $missingSymbols
+            array_diff(
```
Comment (villfa) [inline src/Reflector.php:105]: You should add `use function array_diff` on top of the file.

File: src/Reflector.php [hunk 6]
```diff
@@ -87,18 +87,27 @@ final class Reflector
...
+     * @param array<string, int>        $missingSymbols
      */
     private static function initSymbolList(?array &$symbols, array $source, array $missingSymbols): void
     {
         if (null !== $symbols) {
             return;
         }
 
+        $excludingSymbols = array_keys(
+            array_filter($missingSymbols, static function ($version) {
```
Comment (theofidry) [inline src/Reflector.php:99]: missing imports for:

- `array_filter`
- `PHP_VERSION_ID`

Note that this is configurable on phpStorm if you use it (look for "import function")

Also minor, but I tend to prefer this CS:

```php
array_keys(
    array_filter(
        $missingSymbols,
        static function (int $version): bool {
             return PHP_VERSION_ID < $version;
        }
    )
);
```

**Review by theofidry — APPROVED** [review]
Thanks a lot! This is a big improvement 👏 

Keep in mind that until we change the phpstorm stubs generated map to record the version used there, we are still missing the version for all of those symbols (the one you did is only the "hot patched" symbols). I'm saying this just as a note to be aware, having this is already quite quite helpful

File: src/Reflector.php (hunk 6 above)
Comment (smoench) [reply src/Reflector.php:99]: Next todo: add a linter job for phpcs in GitHub Action 😄

**Comment by smoench** [comment]
> Keep in mind that until we change the phpstorm stubs generated map to record the version used there, we are still missing the version for all of those symbols (the one you did is only the "hot patched" symbols). I'm saying this just as a note to be aware, having this is already quite quite helpful

Yes, I had this in mind. But unfortunately `T_MATCH` has no `@since`-annotation https://github.com/JetBrains/phpstorm-stubs/blame/master/tokenizer/tokenizer.php#L133

File: src/Reflector.php (hunk 5 above)
Comment (smoench) [reply src/Reflector.php:111]: fixed

File: src/Reflector.php (hunk 6 above)
Comment (smoench) [reply src/Reflector.php]: imports and cs fixed

**Comment by smoench** [comment]
Applied proposals, CS and resolved conflicts. Ready for merge :-)

A follow up task would be generating a map of classes, function and constants with the related php version they have been introduced. (or similar)

**Comment by theofidry** [comment]
Cool, thanks a lot for the work @smoench!

**Pull request merged**